File: problems/0235-lowest-common-ancestor-of-a-bst/python/solution.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def from_level(vals):
    if not vals or vals[0] is None:
        return None
    root = TreeNode(vals[0])
    queue = deque([root])
    i = 1
    while queue and i < len(vals):
        node = queue.popleft()
        if i < len(vals) and vals[i] is not None:
            node.left = TreeNode(vals[i])
            queue.append(node.left)
        i += 1
        if i < len(vals) and vals[i] is not None:
            node.right = TreeNode(vals[i])
            queue.append(node.right)
        i += 1
    return root


def lowestCommonAncestor(root, p, q):
    """LCA of a BST — https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/"""
    cur = root
    while cur:
        if p > cur.val and q > cur.val:
            cur = cur.right
        elif p < cur.val and q < cur.val:
            cur = cur.left
        else:
            return cur
    return None


def lcaArr(vals, p, q):
    return lowestCommonAncestor(from_level(vals), p, q).val

File: problems/0235-lowest-common-ancestor-of-a-bst/go/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func fromLevel(vals []*int) *TreeNode {
	if len(vals) == 0 || vals[0] == nil {
		return nil
	}
	root := &TreeNode{Val: *vals[0]}
	queue := []*TreeNode{root}
	i := 1
	for len(queue) > 0 && i < len(vals) {
		node := queue[0]
		queue = queue[1:]
		if i < len(vals) && vals[i] != nil {
			node.Left = &TreeNode{Val: *vals[i]}
			queue = append(queue, node.Left)
		}
		i++
		if i < len(vals) && vals[i] != nil {
			node.Right = &TreeNode{Val: *vals[i]}
			queue = append(queue, node.Right)
		}
		i++
	}
	return root
}

// LCA of a BST — https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
func lowestCommonAncestor(root *TreeNode, p, q int) *TreeNode {
	cur := root
	for cur != nil {
		if p > cur.Val && q > cur.Val {
			cur = cur.Right
		} else if p < cur.Val && q < cur.Val {
			cur = cur.Left
		} else {
			return cur
		}
	}
	return nil
}

func lcaArr(vals []*int, p, q int) int {
	return lowestCommonAncestor(fromLevel(vals), p, q).Val
}
